.github/workflows/prepare.yaml

```
@@ -0,0 +1,28 @@
+name: prepare
+
+on:
+  push:
+    branches: [master]
+    tags-ignore: ["v*"]
+  pull_request:
+    branches: [master]
+
+jobs:
+  main:
+    runs-on: ubuntu-latest
+
+    steps:
+      - name: Clone repository
+        uses: actions/checkout@v3
+        with:
+          fetch-depth: 0
+          submodules: recursive
+
+      - name: Install shellcheck
+        run: sudo apt install shellcheck
+
+      - name: Run tests with BATS
+        run: ./test/bats/bin/bats test/*.bats
+
+      - name: Run shellcheck for changed files
+        run: shellcheck $(git diff --name-only origin/master | grep "bin/*")
```

Three potential problems here:

* `git-diff` will quote/escate paths which contain some special characters and `shellcheck` won't be able to understand it (see `core.quotePath` in `git-config(1)`)
* `$()` will perform word splitting after substitution, which will cause paths with whitespaces to be passed to `shellcheck` as if they were multiple separate files.
* `bin/*` regex will match `bin` in any part of the path, even something like `src/actually-not-binaries-at-all`. Are you sure this is what you want? Assuming it's not, maybe try something like this instead?

```suggestion
        run: git diff --name-only -z origin/master bin/ | xargs -0 shellcheck
```

---

Thanks, very good points.